fix: add HTTPRoute controller for watching HTTPRoute to enqueue KongUpstreamPolicy needing to update ancestor status

randmonkey opened the pull request:
<!-- Thanks for sending a pull request! Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/guide/release-notes.md and ensure your changes are being reflected in CHANGELOG.md for the next upcoming release
-->

**What this PR does / why we need it**:

<!-- Please describe why this particular PR is necessary or why you see it as a nice addition -->
Add a controller for watching `HTTPRoute` to watch `HTTPRoute`s if ancestor statos of related `KongUpstreamPolicy`.
Removes the dependency of `HTTPRoute` of `KongUpstreamPolicy`.

**Which issue this PR fixes**:

<!--
Here you can add any links to issues that this PR is relevant for.
You can use Github keywords (like: closes, fixes or resolves) to auto-resolve
the linked issue(s) when this PR gets merged.

For example: fixes #<issue number>
-->

fixes #5729 

**Special notes for your reviewer**:

<!-- Here you can add any open questions or notes that you might have for reviewers -->

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

czeslavo commented on internal/manager/controllerdef.go:
```diff
@@ -250,11 +260,17 @@ func setupControllers(
+		},
+		// HTTPRoute controller for updating KongUpstreamPolicy's ancestor status with services used as backends of HTTPRoutes.
+		{
+			Enabled: c.KongUpstreamPolicyEnabled && c.GatewayAPIHTTPRouteController,
```
As I understand, if we have Gateway API CRDs installed, this will run two controllers (`kongUpstreamPolicyReconciler` and the decorated one). I think we should make this mutually exclusive - only one of them should be running, otherwise we'll be processing two requests per KUP update.

randmonkey replied:
I added one more controller because I want to start watching `HTTPRoute` when the CRD is installed. If we think it is OK to require restart of KIC to enable watching of `HTTPRoute`, it should be much easier to implement and fewer chances for bugs.

randmonkey replied:
From the test results, I found there is no simple way to enable it to watch `HTTPRoute` dynamically. So I removed the extra controller and added a `CRDExist` check in `KongUpstreamPolicy` controller.

czeslavo commented on internal/manager/controllerdef.go:
```diff
@@ -250,11 +260,17 @@ func setupControllers(
 			Controller: &crds.DynamicCRDController{
 				Manager:          mgr,
-				Log:              ctrl.LoggerFrom(ctx).WithName("controllers").WithName("Dynamic/KongUpstreamPolicy"),
+				Log:              ctrl.LoggerFrom(ctx).WithName("controllers").WithName("Dynamic/Gateway"),
```
```suggestion
				Log:              ctrl.LoggerFrom(ctx).WithName("controllers").WithName("Dynamic/KongUpstreamPolicyWithHTTPRoute"),
```

randmonkey replied:
Removed this controller.

czeslavo reviewed (changes requested):
I think we should add an envtest in which we ensure that KongUpstreamPolicies used with Services and Ingress work as expected when Gateway API CRDs are not installed. I'm afraid this could still be a problem because of this [HTTPRoute list](https://github.com/Kong/kubernetes-ingress-controller/blob/cfcf7b20bbc8a061d2fcc8cd575a2acaf48b6b95/internal/controllers/configuration/kongupstreampolicy_utils.go#L225) in the original reconciler.

randmonkey commented:
> I think we should add an envtest in which we ensure that KongUpstreamPolicies used with Services and Ingress work as expected when Gateway API CRDs are not installed. I'm afraid this could still be a problem because of this [HTTPRoute list](https://github.com/Kong/kubernetes-ingress-controller/blob/cfcf7b20bbc8a061d2fcc8cd575a2acaf48b6b95/internal/controllers/configuration/kongupstreampolicy_utils.go#L225) in the original reconciler.

Removed the extra dynamic controller and added a `CRDExists` check for `HTTPRoute`. When `HTTPRoute` exists, `KongUpstreamPolicy` controller enables features requiring `HTTPRoute`. Also added test cases in envtest to verify that `KongUpstreamPolicy` works both with and without `HTTPRoute`.

czeslavo commented on internal/controllers/configuration/kongupstreampolicy_controller.go:
```diff
@@ -44,6 +44,9 @@ type KongUpstreamPolicyReconciler struct {
 	// status for KongServiceFacades.
 	KongServiceFacadeEnabled bool
+	// HTTPRouteEnabled determines whether the controller should populate the KongUpstreamPolicy's
+	// ancestor status for HTTPRoutes.
```
```suggestion
	// ancestor status for Services used in HTTPRoutes.
```

czeslavo commented on internal/controllers/configuration/kongupstreampolicy_controller.go:
```diff
@@ -261,7 +268,7 @@ func (r *KongUpstreamPolicyReconciler) getUpstreamPoliciesForHTTPRouteServices(c
 		return nil
 	}
-
+	r.Log.Info("Found HTTPRoute", "name", httpRoute.Name, "namespace", httpRoute.Namespace)
```
```suggestion
```
